From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SoSeDiK <mrsosedik@gmail.com>
Date: Thu, 2 May 2024 02:57:50 +0300
Subject: [PATCH] Prevent creative from overriding items

Adds "items.prevent-creative-override" option to prevent creative from overriding items.
The original item is stored as nbt in custom_data component when sending to player and is restored later from nbt.

diff --git a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
index 9d6b2a44cede2337e3a6ac671395c34cbf337aa6..922cc7498b678ce5a4e63b197eca8bd5f13ae352 100644
--- a/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
+++ b/src/main/java/me/sosedik/kiterino/KiterinoConfig.java
@@ -129,4 +129,11 @@ public final class KiterinoConfig {
 	}
 	// Kiterino end - Item Modifiers API
 
+	// Kiterino start - Prevent creative from overriding items
+	public static boolean preventCreativeItemOverride;
+	private static void preventCreativeItemsOverride() {
+		preventCreativeItemOverride = getBoolean(config, "items.prevent-creative-override", true, "Prevent creative from overriding items");
+	}
+	// Kiterino end - Prevent creative from overriding items
+
 }
diff --git a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
index 26573a3df82aecbfc5000e1791462043339281af..fe7e38c610f1f27317879e68e4ea6c85d1d75b3e 100644
--- a/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
+++ b/src/main/java/me/sosedik/kiterino/modifier/item/ItemModifiersHandlerImpl.java
@@ -18,6 +18,8 @@ import net.minecraft.advancements.AdvancementHolder;
 import net.minecraft.advancements.DisplayInfo;
 import net.minecraft.core.NonNullList;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.core.component.DataComponents;
+import net.minecraft.nbt.CompoundTag;
 import net.minecraft.network.protocol.Packet;
 import net.minecraft.network.protocol.game.ClientboundBundlePacket;
 import net.minecraft.network.protocol.game.ClientboundContainerSetContentPacket;
@@ -524,7 +526,21 @@ public class ItemModifiersHandlerImpl extends ItemModifiersHandler {
 		var bukkitItem = ItemModifier.modifyItem(contextBox);
 		if (bukkitItem == null) return null;
 
-		return ItemStack.fromBukkitCopy(bukkitItem);
+		// Kiterino start - Prevent creative from overriding items
+		ItemStack itemStack = ItemStack.fromBukkitCopy(bukkitItem);
+		CraftPlayer player = (CraftPlayer) contextBox.getViewer();
+		if (player != null && me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && player.getGameMode() == org.bukkit.GameMode.CREATIVE) {
+			net.minecraft.world.item.component.CustomData customData = itemStack.get(DataComponents.CUSTOM_DATA);
+			if (customData == null) {
+				customData = net.minecraft.world.item.component.CustomData.of(new CompoundTag());
+				customData.getUnsafe().put("kiterino_og_item", original.saveOptional(player.getHandle().registryAccess()));
+				itemStack.set(DataComponents.CUSTOM_DATA, customData);
+			} else {
+				customData.getUnsafe().put("kiterino_og_item", original.saveOptional(player.getHandle().registryAccess()));
+			}
+		}
+		return itemStack;
+		// Kiterino end - Prevent creative from overriding items
 	}
 
 }
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
index 7424246750d6ceca1acd5d9ebfd48f0d66504c5d..bff5c763cf42a2e3f1cfb497b9ca6fc5d7c9387d 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayerGameMode.java
@@ -106,6 +106,11 @@ public class ServerPlayerGameMode {
     protected void setGameModeForPlayer(GameType gameMode, @Nullable GameType previousGameMode) {
         this.previousGameModeForPlayer = previousGameMode;
         this.gameModeForPlayer = gameMode;
+        // Kiterino start - Prevent creative from overriding items
+        if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && gameMode == GameType.CREATIVE) {
+            this.player.containerMenu.sendAllDataToRemote();
+        }
+        // Kiterino end - Prevent creative from overriding items
         gameMode.updatePlayerAbilities(this.player.getAbilities());
     }
 
diff --git a/src/main/java/net/minecraft/world/item/ItemStack.java b/src/main/java/net/minecraft/world/item/ItemStack.java
index dc6c07b3cce53da0247fa8e2a7dc70d02a55ade4..1b6c2ea094a25bad8382e601825ea8b6bdd4d320 100644
--- a/src/main/java/net/minecraft/world/item/ItemStack.java
+++ b/src/main/java/net/minecraft/world/item/ItemStack.java
@@ -171,6 +171,15 @@ public final class ItemStack implements DataComponentHolder {
                 if (false && !datacomponentpatch.isEmpty()) { // Paper - This is no longer needed with raw NBT being handled in metadata
                     CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
                 }
+                // Kiterino start - Prevent creative from overriding items
+                if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride) {
+                    net.minecraft.world.item.component.CustomData customData = itemstack.components.get(DataComponents.CUSTOM_DATA);
+                    if (customData != null && customData.contains("kiterino_og_item")) {
+                        Tag itemTag = customData.getUnsafe().get("kiterino_og_item");
+                        itemstack = parse(registryfriendlybytebuf.registryAccess(), itemTag).orElse(itemstack);
+                    }
+                }
+                // Kiterino end - Prevent creative from overriding items
                 return itemstack;
                 // CraftBukkit end
             }
@@ -185,6 +194,19 @@ public final class ItemStack implements DataComponentHolder {
                 // itemstack = itemstack.copy();
                 // CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack)); // Paper - This is no longer with raw NBT being handled in metadata
                 // Spigot end
+                // Kiterino start - Prevent creative from overriding items
+                net.minecraft.world.item.component.CustomData customData = null;
+                if (me.sosedik.kiterino.KiterinoConfig.preventCreativeItemOverride && !itemstack.isEmpty() && me.sosedik.kiterino.modifier.item.ItemModifiersHandlerImpl.modifiers.isEmpty()) {
+                    customData = itemstack.get(DataComponents.CUSTOM_DATA);
+                    if (customData == null) {
+                        customData = net.minecraft.world.item.component.CustomData.of(new CompoundTag());
+                        customData.getUnsafe().put("kiterino_og_item", itemstack.saveOptional(registryfriendlybytebuf.registryAccess()));
+                        itemstack.set(DataComponents.CUSTOM_DATA, customData);
+                    } else {
+                        customData.getUnsafe().put("kiterino_og_item", itemstack.saveOptional(registryfriendlybytebuf.registryAccess()));
+                    }
+                }
+                // Kiterino end - Prevent creative from overriding items
                 ITEM_STREAM_CODEC.encode(registryfriendlybytebuf, itemstack.getItemHolder()); // CraftBukkit - decompile error
                 // Paper start - adventure; conditionally render translatable components
                 boolean prev = net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.get();
@@ -193,6 +215,15 @@ public final class ItemStack implements DataComponentHolder {
                 DataComponentPatch.STREAM_CODEC.encode(registryfriendlybytebuf, itemstack.components.asPatch());
                 } finally {
                     net.minecraft.network.chat.ComponentSerialization.DONT_RENDER_TRANSLATABLES.set(prev);
+                    // Kiterino start - Prevent creative from overriding items
+                    if (customData != null) {
+                        if (customData.getUnsafe().size() == 1) {
+                            itemstack.remove(DataComponents.CUSTOM_DATA);
+                        } else {
+                            customData.getUnsafe().remove("kiterino_og_item");
+                        }
+                    }
+                    // Kiterino end - Prevent creative from overriding items
                 }
                 // Paper end - adventure; conditionally render translatable components
             }
